roqoqo/src/operations/mod.rs

```diff
@@ -882,3 +881,191 @@ pub(crate) fn check_valid_mapping(mapping: &HashMap<usize, usize>) -> Result<(),
     }
     Ok(())
 }
+
+/// Represents bosonic modes involved in a roqoqo bosonic Operation.
+#[derive(Debug, PartialEq, Clone, Eq)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub enum InvolvedModes {
+    /// Operation affects all bosonic modes no matter how many there are.
+    All,
+    /// Operation affects no bosonic modes (annotations etc.).
+    None,
+    /// Operation affects a specific set of bosonic modes.
+    Set(HashSet<usize>),
+}
+
+/// Trait for the bosonic modes involved in each bosonic Operation.
+///
+/// # Example
+/// ```
+/// use roqoqo::operations::{InvolveModes, InvolvedModes};
+/// use std::collections::{HashMap, HashSet};
+/// ```
+pub trait InvolveModes {
+    /// Returns all bosonic modes involved in operation.
+    fn involved_modes(&self) -> InvolvedModes;
+}
+
+/// SubstituteModes trait allowing to perform bosonic mode mappings.
+///
+/// # Example
+/// ```
+/// use roqoqo::operations::{SubstituteModes};
+/// use qoqo_calculator::{Calculator, CalculatorFloat};
+/// use std::collections::HashMap;
+///
+/// let mut mode_mapping_test: HashMap<usize, usize> = HashMap::new();
+/// mode_mapping_test.insert(0, 2);
+/// mode_mapping_test.insert(2, 0);
+/// ```
+///
+pub trait SubstituteModes
+where
+    Self: Sized,
+{
+    /// Remaps the bosonic modes in clone of the operation.
+    fn remap_modes(&self, mapping: &HashMap<usize, usize>) -> Result<Self, RoqoqoError>;
+}
+
+/// Trait for Operations acting with a unitary gate on a set of bosonic modes.
+///
+/// # Example
+/// ```
+/// use ndarray::array;
+/// use num_complex::Complex64;
+/// use roqoqo::operations::{OperateModeGate};
+///
+/// let matrix = array![
+///     [Complex64::new(0.0, 0.0), Complex64::new(1.0, 0.0)],
+///     [Complex64::new(1.0, 0.0), Complex64::new(0.0, 0.0)]
+/// ];
+/// ```
+///
+pub trait OperateModeGate:
+    Operate + InvolveModes + SubstituteModes + Clone + PartialEq + SupportedVersion
+{
+    /// Returns unitary matrix of the gate.
+    fn unitary_matrix(&self) -> Result<Array2<Complex64>, RoqoqoError>;
```

Comment:
The general gate matrix would operate on infinite states, so we cannot simply return it like this. We should discuss if we want to return a unitary matrix at all.